libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -71,9 +71,231 @@ const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("ESC_BM", 3, AP_UAVCAN, _esc_bm, 255),
 
+    // @Param: FIX_R
+    // @DisplayName: UAVCAN GNSS Fix broadcast rate
+    // @Description: UAVCAN GNSS Fix broadcast rate in times per second. 0 - disable broadcast.
+    // @Range: 0 20
+    // @User: Advanced
+    AP_GROUPINFO("FIX_R", 4, AP_UAVCAN, _broadcast_fix_rate, 10),
+
+    // @Param: FIX2_R
+    // @DisplayName: UAVCAN GNSS Fix2 broadcast rate
+    // @Description: UAVCAN GNSS Fix2 broadcast rate in times per second. 0 - disable broadcast.
+    // @Range: 0 20
+    // @User: Advanced
+    AP_GROUPINFO("FIX2_R", 5, AP_UAVCAN, _broadcast_fix2_rate, 10),
+
+    // @Param: ATT_R
+    // @DisplayName: UAVCAN attitude broadcast rate
+    // @Description: UAVCAN attitude broadcast rate in times per second. 0 - disable broadcast.
+    // @Range: 0 20
+    // @User: Advanced
+    AP_GROUPINFO("ATT_R", 6, AP_UAVCAN, _broadcast_att_rate, 10),
+
     AP_GROUPEND
 };
 
+// publisher interfaces
+static uavcan::Publisher<uavcan::equipment::gnss::Fix> *fix_out_array[MAX_NUMBER_OF_CAN_DRIVERS];
+static uavcan::Publisher<uavcan::equipment::gnss::Fix2> *fix2_out_array[MAX_NUMBER_OF_CAN_DRIVERS];
+static uavcan::Publisher<uavcan::equipment::ahrs::Solution> *attitude_out_array[MAX_NUMBER_OF_CAN_DRIVERS];
+
+static uavcan::equipment::gnss::Fix _fix_state;
+static uavcan::equipment::gnss::Fix2 _fix2_state;
+static uavcan::equipment::ahrs::Solution _att_state;
+
+void AP_UAVCAN::UAVCAN_AHRS_update(const AP_AHRS_NavEKF &ahrs)
+{
+    bool sem_ret;
+    if ((sem_ret = _fix_out_sem->take(1)) &&
```

Comment:
I get a little lost in this long function. Would you consider changing it to something like:

fix_update()
fix2_update()
ahrs_update()

Reply:
Yes, that's long but no point in separating. It's totally sequential and separation will not change anything.